drivers/photo-electronic/adpd410x/adpd410x.c

```diff
@@ -0,0 +1,660 @@
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include "error.h"
+
+#include "adpd410x.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Read device register.
+ * @param dev - Device handler.
+ * @param address - Register address.
+ * @param data - Pointer to the register value container.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_reg_read(struct adpd410x_dev *dev, uint16_t address,
+			  uint16_t *data)
+{
+	int32_t ret;
+	uint8_t buff[] = {0, 0, 0, 0};
+
+	if(dev->dev_type == ADPD4100) {
+		buff[0] = (address & 0x7f80) >> 7;
+		buff[1] = (address & 0x7f) << 1;
+		buff[1] &= 0xfe;
+
+		ret = spi_write_and_read(dev->dev_ops.spi_phy_dev, buff, 4);
+		if(ret != SUCCESS)
+			return FAILURE;
+	} else {
+		buff[0] = (address & 0x7f00) >> 8;
+		buff[0] |= 0x80;
+		buff[1] = address & 0xff;
+
+		/* No stop bit */
+		ret = i2c_write(dev->dev_ops.i2c_phy_dev, buff, 2, 0);
+		if(ret != SUCCESS)
+			return FAILURE;
+		ret = i2c_read(dev->dev_ops.i2c_phy_dev, (buff + 2), 2, 1);
+		if(ret != SUCCESS)
+			return FAILURE;
+	}
+
+	*data = ((uint16_t)buff[2] << 8) & 0xff00;
```

Comment:
nitpick: I don't think 0xff00 and 0xff masks are needed, those bits will be zero after shifting.

Reply:
I would keep this if it's not too bad. I learned at an university course about architectures that actually duplicated the last bit instead of filling with zero when shifting.